lib/src/source/git.dart

```diff
@@ -196,17 +196,67 @@ class GitSource extends CachedSource {
 
   /// Given a Git repo that contains a pub package, gets the name of the pub
   /// package.
+  ///
+  /// Will download the repo to the system cache under the assumption that the
+  /// package will be downloaded afterwards.
   Future<String> getPackageNameFromRepo(
-      String repo, String? ref, String? path, SystemCache cache) {
-    // Clone the repo to a temp directory.
-    return withTempDir((tempDir) async {
-      await _clone(repo, tempDir, shallow: true);
-      if (ref != null) await _checkOut(tempDir, ref);
-      var pubspec = Pubspec.load(p.join(tempDir, path), cache.sources);
-      return pubspec.name;
+    String url,
+    String? ref,
+    String? path,
+    SystemCache cache, {
+    required String containingDir,
+  }) async {
+    final description = GitDescription(
+      url: url,
+      ref: ref,
+      path: path,
+      containingDir: containingDir,
+    );
+    return await _pool.withResource(() async {
+      await _ensureRepoCache(description, cache);
+      var path = _repoCachePath(description, cache);
+      var revision = await _firstRevision(path, description.ref);
+      final resolvedDescription = GitResolvedDescription(description, revision);
+
+      return Pubspec.parse(
+        await _showFileAtRevision(resolvedDescription, 'pubspec.yaml', cache),
+        cache.sources,
+      ).name;
     });
   }
 
+  /// Lists the file as it is represented at the revision of [description].
+  ///
+  /// Assumes that revision is present in the cache already (can be done with
+  /// [_ensureRevision]).
+  Future<String> _showFileAtRevision(
+    GitResolvedDescription resolvedDescription,
+    String pathInProject,
```

Comment:
Do you want to use this argument? Or rename the method to `pubspecAtRevision`

Reply:
Good catch - yeah, we should use it. copy paste error :)